### ISSUE TITLE
Recommend a workspace-class that is smaller than g1-standard and test feasibility

### ISSUE THREAD

**Issue by kylos101** [gitpod-io/gitpod#12708]
## Is your feature request related to a problem? Please describe

From a harvester preview VM or workspace-preview, we want to see if a small workspace class (smaller than g1-standard) performs well enough for simple workloads. If it is a good experience, then, we'd want to amend webapp and workspace, to support a new "small" workspace class, prior to enabling UBP and workspace classes for individuals.

[Internal context](https://www.notion.so/J-rgen-s-additional-thoughts-on-UBP-free-tier-and-pre-paid-plans-dc5fe164b76e4198acf09e21d58459db?d=15d7d50594c9406581706ff1a29c181c#5b0c6cef072f435b86ad79097c7c8091)
[More internal context](https://gitpod.slack.com/archives/C01KGM9AW4W/p1662494069241589)

cc: @mbrevoort @atduarte 

## Describe the behaviour you'd like

_Idea_: alter `ws-manager` and `ws-daemon` configs, and test a regular workspace with a class using a [limit CPU](https://github.com/gitpod-io/ops/blob/0ed8b6002829035722b57791f12408ee1e33e392/deploy/workspace/installer-values.yaml#L45-L46) of 1 and [burst CPU](https://github.com/gitpod-io/ops/blob/0ed8b6002829035722b57791f12408ee1e33e392/deploy/workspace/installer-values.yaml#L45-L46) of 2 (ws-daemon), memory requests of 2Gi and a limit of 4Gi (kubernetes), with 15Gi of [storage](https://github.com/gitpod-io/ops/blob/0ed8b6002829035722b57791f12408ee1e33e392/deploy/workspace/installer-values.yaml#L178), and 5Gi of [ephemeral storage](https://github.com/gitpod-io/ops/blob/0ed8b6002829035722b57791f12408ee1e33e392/deploy/workspace/installer-values.yaml#L179). Also, please be sure to enable disk IO limiting on the related ws-daemon, this way the workspace is limited similarly to how we would do in production.

Then, once you're able to start workspaces using the above config, test that the `small` workspace class configured is works well enough as a regular workspace to develop [our website](https://github.com/gitpod-io/website). For example:
1. Make change codes
2. How does the file watcher pick them up
3. Run tests
4. Commit code
5. [Use extensions associated with the project](https://github.com/gitpod-io/website/blob/9c0917d838d7ecb7f48162c34521fcba94949eb0/.gitpod.yml#L50-L52)

What about our Gitpod [repo](https://github.com/gitpod-io/gitpod)? How does it behave from a development standpoint? The assumption is that it will not a great experience.

## Describe alternatives you've considered

Some extensions are resource intense. @akosyakov , can you think of any extensions that are common, but hungry, that we should include as part of this test?

## Additional context

Will this be useful enough for JetBrains? @akosyakov wdyt? I assume no, because this workspace will not meet minimum requirements.

## Definition of done
If feasible for users, a smaller workspace-class recommendation is shared and agreed with Product and Finance teams, and related issues are added to groundwork to support the related deployment.

**Comment by mbrevoort** [comment]
Overall we are trying to assess if there is a viable workspace resource size below `g1-standard` that could support development of what type of projects. The resources described above seem like a great place to start.

**Comment by mbrevoort** [comment]
In addition, how would such a resource profile affect pod scheduling, resource utilization and noisy neighbor conditions?

**Comment by akosyakov** [comment]
> Some extensions are resource intense. @akosyakov , can you think of any extensions that are common, but hungry, that we should include as part of this test?

java, golang, typescript. keep in mind hungry depends on size and complexity of the project, not extensions itself

**Comment by Furisto** [comment]
Tested this with the gitpod-io/website repository. Running tests and making code changes works well but trying to build the website leads to OOM kills or network disconnects due to not being enough memory available.

**Comment by kylos101** [comment]
@mbrevoort good questions!

> In addition...

> how would such a resource profile affect pod scheduling

The pods would be scheduled to a dedicated node pool for the small workspace class, so, they wouldn't impact pods being scheduled to other node pools (like standard and large). 

Initially we'd control density via memory requests, but, may want to consider CPU requests, too. For example, memory requests would be 2Gi, and CPU might be 1 core, or .5 cores, etc.

I do wonder how many small workspaces we could fit on a node? But, first we need to find a size that works for one workspace. And then we can talk about achieving a desired density. For example, if we run more than 18 small workspaces, perhaps 36, we'd have to reduce the related disk IO bandwidth.

> resource utilization 

The way this is written now, you'd get 1 CPU, and be able to burst to 2 (controlled by ws-daemon), and you wouldn't be able to use more than 4Gi of memory (controlled by Kubernetes).

> and noisy neighbor conditions?

@Furisto can you think of anything new that we might have from a risk perspective, by having a higher density of workspaces?

**Comment by kylos101** [comment]
> Tested this with the gitpod-io/website repository. Running tests and making code changes works well but trying to build the website leads to OOM kills or network disconnects due to not being enough memory available.

Okay, good to know, @Furisto !  Bummed to hear, though. :wink:

**Comment by Furisto** [comment]
> @Furisto can you think of anything new that we might have from a risk perspective, by having a higher density of workspaces?

Higher chance of WorkspaceStuckInStopping alerts because more workspaces mean more backups and we have a concurrency limit for backups. On the other hand the backups should be smaller, so they will go faster.

**Comment by kylos101** [comment]
@Furisto can you share a link to the branch where you made these related changes? It would be good to get a second pair of eyes 👀  on the related configuration for the experiment.

@mbrevoort would you like any further investigation to this resource permutation, or other resource permutations? The small configuration we originally socialized does not look promising.

**Comment by jldec** [comment]
> Tested this with the gitpod-io/website repository. Running tests and making code changes works well but trying to build the website leads to OOM kills or network disconnects due to not being enough memory available.

@Furisto @kylos101 - Would you be able to share more details about which processes are consuming more memory and causing problems when building the website?

**Comment by kylos101** [comment]
👋 @Furisto in hindsight, in talking with @aledbf , let's change the test as follows:

1. Make a new branch in the ops repo with these related changes, and share with us here for reference
2. Deploy an ephemeral cluster
3. Share cluster access using `new-workspace-cluster` to @mbrevoort and @jldec , so they can experience the related performance/behavior.

**Comment by Furisto** [comment]
@kylos101 @jldec Tested this on an ephemeral cluster and the performance looks better there. See https://www.loom.com/share/fc6fb07dd05b4621841d90f5a0f41dc8. I have given you access to the ephemeral cluster @jldec

**Comment by kylos101** [comment]
@jldec let us know what you think? 🙏 

Wow, nice, @Furisto ! As a next step, I recommend you:
1. Share a link to the branch containing your configuration changes? This way we can review it.
2. Do a test with loadgen and a regular workspace you actively use. For example, fill up a node (leaving enough room for 1 regular workspace), and then connect with a regular workspace, to see what the performance is like? I ask because of related limiting/tuning that could be needed.
3. Think whether any other tests are needed? 🤔 

Depending on output from you both, we'll need to create some issues:
1. add the class configuration to webapp and workspace
2. update logic in server, so we can have two pools, one for PVC and one without
3. update configuration in the packer image, to account for the new pools
4. update configuration in the ops repo, to build the related instance groups

CC: @atduarte for awareness

**Comment by jldec** [comment]
Initial (single workspace) evaluation with the gitpod-io/website repo did not reveal any issues.
I was able to `npm run  build` and edit content while watching the dev server live-reload.
Memory and CPU both crept up toward 100% during the build, but I did not observe any errors.

**Comment by Furisto** [comment]
This is the configuration that I am using. I just edited the configmap of ws-manager directly:
```
"g1-standard": {
            "name": "",
            "container": {
              "requests": {
                "cpu": "1m",
                "memory": "3328Mi",
                "ephemeral-storage": "5Gi"
              },
              "limits": {
                "cpu": {
                  "min": "1",
                  "burst": "2"
                },
                "memory": "4Gi",
                "ephemeral-storage": "5Gi",
                "storage": "15Gi"
              }
            },
            "templates": {
              "defaultPath": "/workspace-templates/g1-standard-default.yaml",
              "regularPath": "/workspace-templates/g1-standard-regular.yaml",
              "prebuildPath": "/workspace-templates/g1-standard-prebuild.yaml",
              "imagebuildPath": "/workspace-templates/g1-standard-imagebuild.yaml"
            },
            "pvc": {
              "size": "15Gi",
              "storageClass": "csi-gce-pd-g1-standard",
              "snapshotClass": "csi-gce-pd-snapshot-class"
            }
          },
```
Here [is a video](https://www.loom.com/share/49369e974020492bbf2f0c2d7e610bae) of me using the workspace while the loadtest is running. I am building the project while editing the code and just navigating around. You can see that code completion is not working very well, but otherwise it is fine. Once the build was complete, code completion was ok again.

**Comment by Furisto** [comment]
@jldec @mbrevoort @atduarte Given the above information, would you like to proceed with adding a smaller workspace class?

**Comment by kylos101** [comment]
@Furisto thank you for this bit of info:
> You can see that code completion is not working very well, but otherwise it is fine. Once the build was complete, code completion was ok again.

I think that is acceptable given the small workspace class. Appreciate you sharing the result! :+1:

**Comment by jldec** [comment]
@kylos101 - I did not see the same long delays as in the video when I tested the website repo with the small workspaces myself. Maybe I had a better connection to the environment (I was in Dublin) or maybe the editor caches in my workspace were more warmed up. 

I think another test would be helpful to better understand the behavior, but in general I would not hold up the introduction of small workspaces for this reason alone.
